File: src/fragment/BlockTypeFragment.ts
import { HTMLText, Text } from "pixi.js";
import Fragment, { FragmentType, register } from "./Fragment";
import { StructEndecBuilder } from "KEndec";
import { Identifier } from "~/util";


export default class BlockTypeFragment extends Fragment {

    blockType: string;

    constructor(blockType: string) {
        super();

        this.blockType = blockType;
    }

    override toString(): string {
        return this.blockType
    }

    override type(): FragmentType<BlockTypeFragment> {
        return BLOCK_TYPE;
    }
}

const BLOCK_TYPE = register("block_type", 0x44aa33, 
    StructEndecBuilder.of1(
        Identifier.ENDEC
            .xmap((ident) => ident.toString(), Identifier.of)
            .fieldOf("block", (fragment: BlockTypeFragment) => fragment.blockType),
        (str) => new BlockTypeFragment(str) 
    )
);

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/BlockTypeFragment.kt
package me.maplesyrum.tricksterstudio.spell.fragment

import Identifier
import me.maplesyrum.tricksterstudio.external.pixi.HTMLText
import tree.maple.kendec.impl.StructEndecBuilder


class BlockTypeFragment(
    val blockType: Identifier
) : Fragment() {
    override fun type(): FragmentType<*> = FragmentType.BLOCK_TYPE

    companion object {
        val ENDEC = StructEndecBuilder.of(
            Identifier.ENDEC.fieldOf("block", BlockTypeFragment::blockType),
            ::BlockTypeFragment
        )
    }

    override fun toString(): String = blockType.toString()
}